tx relay v2

0xFFFC0000 opened the pull request:
https://github.com/monero-project/monero/pull/9933

0xFFFC0000 committed:
Tx Relay v1: New Relay Logic with threshold-based peer dropping.

Co-authored-by: Boog900 <boog900@tutanota.com>

0xFFFC0000 committed:
tx-relay-v2:
* Removed extra classes.
* Used on_idle as callback mechanism.

0xFFFC0000 committed:
Fixed extra delete directives. Only delete methods
that compiler generates implicitly.

0xFFFC0000 committed:
tx-relay-v2: add maximum in_flight requests for each connection

0xFFFC0000 committed:
change internal txrequest representation to multiindex hash map

0xFFFC0000 committed:
Addressing second round of comments

nahuhh commented:
in/out is in MiB

| run | in | out | total (1hr) | time | version info | exclusive node |
| --- | --- |--- | --- | --- | --- | --- |
| 1 | 42.95 | 18.37 | 61.32 (29.20) | 2.1hr | 1.2 branch (incl 177) + relayv2 | same
| 1 | 123.24 | 1.72 | 124.96 (59.50) | 2.1hr | 1.2 branch (incl 177) | 1.2 release |
| 2 | 89.68 | 15.06 | 104.75 (20.54) | 5.1hr | 1.2 release + relayv2 | same |
| 2 | 783.83 | 4.85 | 788.68 (154.64) | 5.1hr | 1.2 branch (incl 177) | 1.2 release |
| 3 | 47.51 | 7.71 | 52.22 (22.09) | 2.5hr | 1.2 release + relayv2 | same
| 3 | 713.88 | 6.72 | 723.60 (289.44) | 2.5hr | 1.2 release | same

observations:
- tx relay v2 seems to increase uploaded data, but _significantly_ reduces downloaded and total data.
- tx-relay on its own used about 10-30% as much data as 177 w/o v2.
- baseline vs 177: baseline used ~2x as much total data as 177. (different runs. Didnt test side-by-side)
- checked tx pools at the end if the tests (they match)
- ~~note: have not yet tested a baseline of 1.2 release (w/o v2 and 177)~~

0xFFFC0000 committed:
fix 9933#discussion_r2366464294

0xFFFC0000 committed:
fix for 9933#discussion_r2366462339 and 9933#discussion_r2366458244

0xFFFC0000 committed:
fix for 9933#discussion_r2366295474

0xFFFC0000 committed:
fix for 9933#discussion_r2366482602

0xFFFC0000 committed:
fix for 9933#discussion_r2366489004

0xFFFC0000 committed:
fix for 9933#discussion_r2366493055

0xFFFC0000 commented:
1. Rebased to latest [fcmp++-alpha-stressnet](https://github.com/seraphis-migration/monero/tree/fcmp%2B%2B-alpha-stressnet).
2. Includes the latest fixes from: https://github.com/monero-project/monero/pull/9933/commits

0xFFFC0000 committed:
improvements, fix the deadline logic, better logging

SChernykh commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -1786,6 +1946,75 @@ skip:
+        if (req.is_in_flight())
+          {
+            bool drop_peer = false;
+            if (!req.is_marked_for_removal() && elapsed.count() > deadline)
```
`elapsed` counts time since first seen, not since `fly()` was called, so it can timeout 1 millisecond after the `fly()` call. Is it indended?

0xFFFC0000 replied:
That is request callback. Gets called in predefined intervals. Yes that is intended behaviour.

SChernykh commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -953,25 +1064,42 @@ namespace cryptonote
+            && m_request_manager.already_requested_tx(tx_hash))
+        {
+          m_request_manager.received_requested(context.m_connection_id, tx_hash);
+          context.remove_in_flight_request();
```
You don't check the return value of `received_requested` here. If it returns `false`, then tx_hash wasn't received and you shouldn't remove in flight request, no?

0xFFFC0000 replied:
I don't understand your point. At this point we already received the transaction. 

The only issue here is when the received tx is not the one we requested from a connection id. Which is a known issue and will be addressed in next update.

SChernykh commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -902,6 +933,86 @@ namespace cryptonote
+      if (!context.can_process_additional_request())
+      {
+        LOG_ERROR_CCONTEXT("Too many in-flight requests, cannot accept");
+        return 1;
```
What's the error return value for this function? It returns 1 on all code paths, is it a logical bug here? Also, it just silently drops all tx hashes once the limit is reached. Maybe it makes sense to queue them for later?

0xFFFC0000 replied:
That m_p2p convention. Nothing in my power.

SChernykh commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -1584,6 +1712,37 @@ namespace cryptonote
+                    if (m_request_manager.get_current_request_peer_id(h, owner) && !owner.is_nil())
+                    {
+                        m_p2p->for_connection(owner, [&](cryptonote_connection_context &ctx, nodetool::peerid_type, uint32_t)->bool {
+                            ctx.remove_in_flight_request();
```
This abandons the in flight request if we received this tx from a block, right? What if this peer already sent this tx to us, and the response will arrive next millisecond? Will it call `remove_flight_request` again? But I guess it's fine because you check if it's `> 0` there.

0xFFFC0000 replied:
It will be treated like what we used to do.

SChernykh commented on src/cryptonote_basic/connection_context.cpp:
```diff
@@ -89,4 +96,156 @@ namespace cryptonote
+    MINFO("Peer " << epee::string_tools::pod_to_hex(m_connection_id)
+          << " has missed " << m_connection_stats->missed << " out of "
+          << announced << " announced txs (" << percent << "%)");
+    return percent > P2P_REQUEST_FAILURE_THRESHOLD_PERCENTAGE;
```
This should have a minimum sample size. Right now, if a peer misses its first request, it gets 100% miss ratio and then gets dropped. Not fair, I think.

0xFFFC0000 replied:
Good point. I will add that.

0xFFFC0000 replied:
Fixed it, thanks.

SChernykh commented on src/cryptonote_protocol/request_manager.cpp:
```diff
@@ -0,0 +1,255 @@
+  }
+
+  // Recompute range after potential erasures.
+  by_tx_hash = get_requests_by_tx_hash(m_requested_txs);
```
I don't understand what's going on at this line. `by_tx_hash` is a reference to something, are you reassigning it? It's an unconventional use of C++ references, so I'm confused. Maybe it's better to give it a new name `by_tx_hash2` or similar to not mix with the above code?

0xFFFC0000 replied:
This part is going to get removed. Nothing important.

SChernykh commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -902,6 +933,86 @@ namespace cryptonote
+    std::vector<blobdata> txs;
+
+    // Iterate over requested txin hashes
+    for (const auto &tx_hash : arg.t)
```
This doesn't check for max in flight requests limit, so a malicious peer can send a max size packet and squeeze as many tx hashes as possible into it. They can even pick a single tx hash of some "fat" transaction currently in the mempool and make the reply very big.

Edit: an attacker can request the same hash N times (as many times as possible) - the hash of the biggest transaction in the pool. The code doesn't check for duplicate hashes.

Edit 2: 2 MB packet can ask for 65536 tx hashes, and if they all ask for the same 150 KB transaction (max size allowed in the mempool), this is 9.15 GB of data in the `txs` vector. If many peers do this at the same time, they for sure will DoS the node.

j-berman replied:
Agree with @SChernykh, would be good to check for duplicates here

SChernykh commented on src/cryptonote_basic/connection_context.cpp:
```diff
@@ -89,4 +96,156 @@ namespace cryptonote
+  {
+    MINFO("Decrementing in_flight_requests count for peer: " << epee::string_tools::pod_to_hex(m_connection_id) << ", current in_flight_requests: " << m_connection_stats->in_flight_requests);
+    std::unique_lock<std::shared_timed_mutex> w_lock(m_connection_stats->mutex);
+    if (m_connection_stats->in_flight_requests > 0)
```
This check means that `in_flight_requests` is not an exact counter and we can underestimate how many request this peer actually has?

0xFFFC0000 replied:
It is roughly exact number. But we should have defense against any network inconsistency in all layers. 

The else clause is basically an assert log error message.

SChernykh commented on src/cryptonote_basic/connection_context.cpp:
```diff
@@ -89,4 +96,156 @@ namespace cryptonote
+    ++m_connection_stats->sent;
+  }
+
+  void cryptonote_connection_context::add_in_flight_requests()
```
This function is not used anywhere?

0xFFFC0000 replied:
I am in process of changing this class design a little bit. So this either will be removed or will have another design.

SChernykh commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -902,6 +933,86 @@ namespace cryptonote
+      if (m_core.get_pool_transaction(tx_hash, tx_blob, cryptonote::relay_category::broadcasted))
+      {
+        context.add_sent();
+        txs.push_back(std::move(tx_blob));
```
Should return with an error as soon as `txs` vector gets bigger than the max packet size (100 MB?). This is in addition to the duplicate hash checking.

0xFFFC0000 replied:
100MB? That is extremely extremely huge. Many other limitations will kick in before even getting close to that limitation.

SChernykh replied:
Read the comment above. This array can grow to over 9 GB if an attacker sends a specially crafted request.

0xFFFC0000 replied:
Keep in mind we haven't discussed the maximum number txs that can be requested. We will add a layer of defense to this code.

0xFFFC0000 replied:
Fixed it, thanks.

SChernykh commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -953,25 +1064,42 @@ namespace cryptonote
+      else
+      {
+        context.add_received();
+        // This is extremely inefficient. We have to change m_core.handle_incoming_tx
```
Is it very hard to change `handle_incoming_tx` and make it return the tx hash as an optional output parameter?

0xFFFC0000 replied:
That is not my API, I cannot change the design how p2p interacts.

0xFFFC0000 replied:
That requires a complete change to one of the core APIs of p2p, which is out of scope of this PR.

j-berman replied:
I think it's not a big deal to add a new return by ref for the tx hash in `handle_incoming_tx`, but not a requirement for this PR. Can be handled in a follow-up

SChernykh commented on src/cryptonote_protocol/request_manager.cpp:
```diff
@@ -0,0 +1,255 @@
+        << " the transaction: "
+        << epee::string_tools::pod_to_hex(tx_hash));
+  std::unique_lock<std::recursive_mutex> lock(m_mutex);
+  // Search directly in m_requested_txs instead of using the index
```
Why no index here?

0xFFFC0000 replied:
Read boost multi-index document please.

SChernykh replied:
You could at least search the index first to see if you need to run the loop at all.

0xFFFC0000 replied:
Looking up index via multi_index is not O(1). It is amortized O(n), which means probabilistically it is O(n) 50% of the time. 

Easiest and efficient way to do is a simple go over element.

0xFFFC0000 committed:
new sets of fixes and improvements

j-berman commented on src/cryptonote_basic/connection_context.cpp:
```diff
@@ -89,4 +96,152 @@ namespace cryptonote
     return m_expected_heights[difference];
   }
+
+  void cryptonote_connection_context::reset()
```
unused function

j-berman commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -902,6 +933,92 @@ namespace cryptonote
+
+    std::vector<blobdata> txs;
+
+    if (txs.size() > P2P_MAX_IN_FLIGHT_REQUESTS)
```
```suggestion
    if (arg.t.size() > P2P_MAX_IN_FLIGHT_REQUESTS)
```

j-berman commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -902,6 +933,92 @@ namespace cryptonote
+    {
+      LOG_ERROR_CCONTEXT("Too many in-flight requests, cannot accept request to send " << txs.size() << " txs to " << context.m_connection_id);
+      return 1;
+    }
```
Looks like you meant to use `arg.t.size()` here

0xFFFC0000 replied:
Thanks. Good catch.

j-berman commented on src/cryptonote_basic/connection_context.cpp:
```diff
@@ -89,4 +96,152 @@ namespace cryptonote
+        << ", Sent: " << m_connection_stats->sent
+        << ", Missed: " << m_connection_stats->missed;
+    return oss.str();
+  }
```
These functions are all unused (lines 165 to 229)

0xFFFC0000 replied:
Unused right now. But I want to extend the `connection_context` statistics. I will remove it for now.

j-berman commented on src/cryptonote_protocol/request_manager.cpp:
```diff
@@ -0,0 +1,260 @@
+      }
+  }
+  MINFO("Removed " << removed_count << " requests.");
+}
```
newline

0xFFFC0000 replied:
Thanks, will be fixed in new push.

j-berman commented on src/cryptonote_basic/connection_context.cpp:
```diff
@@ -65,6 +68,10 @@ namespace cryptonote
       return 1024 * 1024; // 1 MB
     case cryptonote::NOTIFY_GET_TXPOOL_COMPLEMENT::ID:
       return 1024 * 1024 * 4; // 4 MB
+    case cryptonote::NOTIFY_TX_POOL_INV::ID:
```
nit: I think `NOTIFY_TX_POOL_HASHES` would work a bit better

0xFFFC0000 replied:
Will change this, thanks.

j-berman commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -128,6 +128,11 @@ namespace cryptonote
   };
+  const command_line::arg_descriptor<size_t> arg_request_deadline  = {
+    "request-deadline"
+  , "Set the deadline for waiting requested transactions (in seconds)"
```
```suggestion
  , "Set the deadline for waiting for requested transactions (in seconds)"
```

0xFFFC0000 replied:
Thanks.

j-berman commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -186,8 +189,10 @@ namespace cryptonote
+                                                                                                              m_no_sync(false),
+                                                                                                              m_request_manager(),
+                                                                                                              m_peer_request_interval([this]() -> uint64_t {return m_interval_peer_request_checker.load(std::memory_order_relaxed);}),
+                                                                                                              m_peer_request_checker(m_peer_request_interval)
```
nit: it looks like this is just going to use the default P2P_DEFAULT_REQUEST_TIMEOUT for `m_peer_request_interval` in every case because it'll call the function in the line above immediately before init() runs, which is fine, but looks unintentional and confusing

0xFFFC0000 replied:
This is another ugly part of the code.

0xFFFC0000 replied:
Will try to find a better solution. sadly it is not possible to change the interval in our idle timer caller/scheduler.

0xFFFC0000 replied:
> nit: it looks like this is just going to use the default P2P_DEFAULT_REQUEST_TIMEOUT for `m_peer_request_interval` in every case because it'll call the function in the line above immediately before init() runs, which is fine, but looks unintentional and confusing

This is not correct. We are able to change `m_peer_request_interval` dynamically later.

j-berman commented on src/cryptonote_config.h:
```diff
@@ -145,6 +145,7 @@
 #define P2P_DEFAULT_PEERS_IN_HANDSHAKE                  250
 #define P2P_MAX_PEERS_IN_HANDSHAKE                      250
 #define P2P_DEFAULT_CONNECTION_TIMEOUT                  5000       //5 seconds
+#define P2P_DEFAULT_REQUEST_TIMEOUT                     (P2P_DEFAULT_CONNECTION_TIMEOUT/1000) // 5 seconds
```
5s seems like not a lot of time, potentially a cause of uptick in bans when running tx relay v2.. log level 1 filled with `Missed tx announcement more that threshold of the time, dropping peer` would indicate that is the case

E.g. let's say you have 5 128-input txs to receive.. even if you assume 0 latency to receive the tx after requesting it, all the calls to `handle_incoming_tx` could take at least 15s to verify the last tx in that batch and mark it received

It may even make sense to avoid calling `check_tx_request_queue` if `handle_notify_new_transactions` is executing, or synchronize the 2 functions with a lock

j-berman commented on src/cryptonote_protocol/levin_notify.cpp:
```diff
@@ -202,12 +180,30 @@ namespace levin
       return out;
     }
 
+    epee::levin::message_writer make_tx_inv_message(std::vector<crypto::hash>&& txs, const bool pad)
```
Unused `pad` variable

0xFFFC0000 replied:
Leftover from when we had pad variable.

j-berman replied:
EDIT: sorry, I think I see what you're saying (that it's a leftover from an earlier state of the PR and will be removed). Can ignore below.

This is a new function this PR introduces that doesn't need `pad` like the other function in the code base does

j-berman commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -902,6 +933,92 @@ namespace cryptonote
+      // If we have the tx in our pool, also remove it from request manager and skip.
+      if (m_core.pool_has_tx(tx_hash))
+      {
+        m_request_manager.remove_announcement(tx_hash);
```
If it's a tx we requested from this peer, looks like it'll still remain in the context's `tx_announcements` as well.

In fact the context's `tx_announcements` struct doesn't shrink anywhere, it will grow forever.

I think there is going to be some trouble keeping these 2 distinct data structures in sync perfectly (the context's stats <> `m_request_manager`). It would be nice to consolidate the logic into one, e.g. maybe move all logic into the `m_request_manager`

j-berman commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -902,6 +933,92 @@ namespace cryptonote
+        continue;
+      }
+      // If we have the tx in our pool, also remove it from request manager and skip.
+      if (m_core.pool_has_tx(tx_hash))
```
Check if in the chain too

0xFFFC0000 replied:
Great point. Thanks. will be addressed in the new push.

j-berman commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -902,6 +933,92 @@ namespace cryptonote
+      if (need_request) {
+        context.add_requested_from_peer();
+        missing_tx_hashes.push_back(tx_hash);
+        m_request_manager.add_request(tx_hash, context.m_connection_id);
```
Doesn't look like you need both `add_announcement` and `add_request`, and can just consolidate the logic of `add_request` into `add_announcement`

0xFFFC0000 replied:
This part is ugly. I am thinking about how to do the logic between `add_announcement` and `add_request`.

j-berman commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -902,6 +933,92 @@ namespace cryptonote
+    // Iterate over each advertised transaction hash and check our pool and our requested tracker.
+    for (const auto &tx_hash : arg.t)
+    {
+      if (!context.can_process_additional_request())
```
P2P_MAX_IN_FLIGHT_REQUESTS=100 seems pretty low to just stop handling new tx hashes and silently ignore them. The daemon wouldn't be able to handle more than 100 new tx hashes it hasn't seen before per connection. I think instead it would make more sense to queue tx hashes for requesting later on... and the `check_tx_request_queue` could be altered to kick off requests that haven't been kicked off before (edit: in addition to the deadline/stale check)

j-berman reviewed (changes requested):
I think this looks pretty good!

Biggest comments:

1) [untested but this is testable] The increasing bans *may*  be caused by handling `check_tx_request_queue` before `handle_notify_new_transactions` can complete. I think it makes sense to grab a lock during `handle_notify_new_transactions` that prevents `check_tx_request_queue`  from executing at the same time.

2) I think `handle_notify_tx_pool_inv` should queue new tx hashes past the in-flight limit to be requested later on rather than silently drop them.

3) To avoid some mistakes I noticed in the code, I would recommend consolidating the context `connection_statistics` into the `m_request_manager`. That way we don't have to make sure to make the calls across both context and `m_request_manager` to make sure they're kept in sync.

0xFFFC0000 commented:
> 3\. To avoid some mistakes I noticed in the code, I would recommend consolidating the context `connection_statistics` into the `m_request_manager`. That way we don't have to make sure to make the calls across both context and `m_request_manager` to make sure they're kept in sync.

I implemented at first like this. But this is the exact opposite of what vtnerd asked me.

j-berman commented:
> I implemented at first like this. But this is the exact opposite of what vtnerd asked me.

His points made sense. Another option would be to pass the connection context as a param into the request manager's functions and modify the relevant stats that way, rather than do it all externally in different places. Example: if you're going to remove an announcement from the request manager, you also want to remove it from the connection's tracking stats. These are inextricably linked logical actions, so consolidating the logic into one place can help avoid bugs like forgetting to do both.

Basically every function in the request manager and new function related to these new connection stats seems to have a corresponding logical tie.

0xFFFC0000 commented:
> His points made sense. Another option would be to pass the connection context as a param into the request manager's functions and modify the relevant stats that way, rather than do it all externally in different places. Example: if you're going to remove an announcement from the request manager, you also want to remove it from the connection's tracking stats. These are inextricably linked logical actions, so consolidating the logic into one place can help avoid bugs like forgetting to do both.

I should think about this. This mean a complete rewrite of 60-70% of the code almost.

0xFFFC0000 committed:
addressing first round of jberman comments

j-berman commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -902,9 +933,96 @@ namespace cryptonote
+
+    std::vector<blobdata> txs;
+
+    if (txs.size() > P2P_MAX_IN_FLIGHT_REQUESTS)
```
`arg.t.size()`

j-berman committed:
Bulk changes for tx relay v2

- Only requests a specific tx hash from a single peer at a time.
- When we receive a tx hash not yet known to us, but connection is at its in-flight limit, we queue it to request later.
  - The tx request checker runs in a 5s interval now, faster than the timeout, in order to fire off these requests in the queue that we're able to (see `fly_available_requests`).
- We never request more than the `max_n_txs_per_packet` at a time from a connection, avoiding connection drops from doing so.
- Handles adding a request and all associated metadata with one `add_request` function.
- Makes sure to remove tx requests from all peers for a specific tx hash as soon as we add the tx to our db.
- Handles removing a request and all associated metadata with one `remove_request` function.
- Modifies `handle_incoming_tx` to return the calculated tx hash.
  - This avoids a circumstance where the node fails to verify a tx with something that's a no-drop-offense (i.e. an expected failure), and we don't mark the tx request as received.
  - Also avoids the extra parsing & hashing.
- Deletes lots of unused logic/metadata/functions.
- Places `m_connection_stats` back in the request manager.
  - I think this initial decision by @0xff was correct.
  - This drastically simplifies synchronized updating of stats relevant to the request manager, and clearly demonstrates what logic the request manager needs to track.
  - If `remove_peer` does not execute upon dropping a peer, then the request data won't get cleaned up also. So we make sure to clean up in the same place.
  - The extra uuid mapping is insignificant on memory.
- Pass tx_hashes directly into send_txs/relay_txs.
- Add unit tests for the request manager.

Boog900 commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
Should this request now respond with `NOTIFY_TX_POOL_HASH`? 

Also, although not done in this PR, I don't _really_ like the idea of responding to requests in multiple messages if the messages trip the packet limit. There are limits on the size of the "pending to send" queue which could trip if multiple big response messages are sent in quick succession. I also think it is a DoS vector. 

The alternative, just truncating the response, isn't great, however I do think it is preferable.

Boog900 commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -1815,6 +1917,80 @@ skip:
+
+    // If we're not synchronized, we shouldn't be requesting any txs. Syncing might end up removing many tx requests
+    // because the txs enter the chain.
+    if (!is_synchronized() || (m_core.get_target_blockchain_height() > m_core.get_current_blockchain_height()))
```
I don't think we should have the `|| (m_core.get_target_blockchain_height() > m_core.get_current_blockchain_height())` here.

A node could fake being ahead of the chain and as `get_target_blockchain_height` can be set higher after you have already synced they can prevent this from running until the request for blocks from them timesout.

`is_synchronized` only updates once, when we think we have reached the top of the chain, after that it wont turn `false` again.

j-berman replied:
My main concern is that if a node falls behind for whatever reason, it can end up missing a number of tx reqs that are already in the chain (other nodes won't respond to our node's tx reqs because they're in the chain already), causing connection drops while syncing.

I think you provide a good reason for removing that || as is, but I also do think we'll want to monitor this issue and prioritize fixing it if it arises.

Perhaps it makes sense to include logic: if our node falls behind and resumes the sync download, clear the tx requests queue.

Boog900 replied:
Ah, so like receiving tx hashes from nodes that are ahead so we don't get the block notifications but also don't get the tx as it is now in the chain.

We probably should just ignore tx hashes on a per-node bases if the node claims to be too far ahead of our chain, this is what Cuprate does. Also in `handle_notify_tx_pool_hash` we should check if `is_synchronized` is false and ignoring the whole message anyway.

That should help, right?

Boog900 commented on src/cryptonote_config.h:
```diff
@@ -154,14 +155,17 @@
 #define P2P_DEFAULT_SYNC_SEARCH_CONNECTIONS_COUNT       2
 #define P2P_DEFAULT_LIMIT_RATE_UP                       8192       // kB/s
 #define P2P_DEFAULT_LIMIT_RATE_DOWN                     32768       // kB/s
+#define P2P_REQUEST_FAILURE_THRESHOLD_PERCENTAGE        70          // if more than 70% of requests fail, the peer is dropped
```
I know I suggested this check to prevent a malicious node from just not relaying all txs, but I do wonder if this could lead to a similar issue like this: https://github.com/monero-project/monero/issues/9335 if the tx-pool is dropping txs too fast.

I don't really know a good solution, maybe we should check txs are still in the pool when the fluff timer fires, as then we reduce the time where a tx can be received then removed while still being sent out? It wouldn't fix this though. 

I don't really think a fix is needed in this PR however I did just want to mention this.

j-berman replied:
> maybe we should check txs are still in the pool when the fluff timer fires, as then we reduce the time where a tx can be received then removed while still being sent out?

Can you expand on this case?

The one case I can think of where the pool may be dropping txs is if the pool is at capacity (max txpool weight) and higher fee txs are displacing the lower fee ones.

I also don't see a great solution to this causing over-exuberant dropped connections.

This general issue is at least mitigated by this being a failure threshold % of total reqs i.e. success cases should strongly outweigh failures especially as time goes on.

Boog900 replied:
> Can you expand on this case?

This wasn't another case. The way the current code works is once a tx is received it is added to the pool and the fluff queue, IIRC the fluffy queue is average of 2.5 seconds for outbound connections and 5 for inbound. If the tx in the pool is then removed before the fluff timer fires we will tell a peer we have a tx when we don't after the timer fires.

If we instead, after the fluff timer fires, check if we still have the tx before sending it, then the time that the tx can be removed while we still tell the peer with have it is reduced. It's not a fix and its also quite heavy handed, it wouldn't be cheap to check the pool after every fluff fire I wouldn't think.

Thinking about it, a better solution might be to not have a queue and to just ask the tx pool for txs to fluff _after_ the timer fires. Just a thought though. 

> This general issue is at least mitigated by this being a failure threshold % of total reqs i.e. success cases should strongly outweigh failures especially as time goes on.

Yeah I agree, hopefully this shouldn't be an issue.

j-berman replied:
> a better solution might be to not have a queue and to just ask the tx pool for txs to fluff after the timer fires

This sounds like a good idea to me for a future PR

j-berman committed:
@boog900 review comments

j-berman committed:
@boog900 review rd2

0xFFFC0000 committed:
Merge pull request #63 from j-berman/stressnet/0xfffc/tx-relay-v2

@boog900 review comments
